Hypershift live migration (#3478)
ovn-org/ovn-kubernetes: docs/design/live_migration.md

# KubeVirt virtual machine bridge binding pod network seamless live migration

## Introduction

At kubevirt every VM is executed inside a "virt-launcher" pod; how the "virt-launcher" pod net interface is "bound" to the VM is specified by the network binding at the VM spec, ovn-kubernetes support live migration of the KubeVirt network bridge binding. 

> At KubeVirt bridge binding the virt-launcher interface is replaced by a bridge with one leg on the pod interface and the other at the VM tap device. When IPAM is configured on the pod interface, a DHCP server is started to advertise the pod's IP to DHCP aware VMs.

Reviewer: This sentence is very confusing.

What are you replacing ? And with what ?...

AFAIU, KubeVirt doesn't replace anything. It adds a bridge and a tap device, and connects the existing pod interface as a port of the aforementioned bridge.

> Benefict of the bridge binding is that is able to expose the pod IP to the VM as expected by most users also pod network bridge binding live migration is implemented by ovn-kubernetes by doing the following:

Reviewer: benefit.

- Do IP assignemnt with ovn-k but skip the CNI part that configure it at pod's netns veth.
- Do not expect the IP to be on the pod netns.
- Add the ability at ovn-k controllers to migrate pod's IP from node to node.

## Implementation 

To implement live migration ovn-kubernetes do the following:

> - Deliver IP address at VMs configuring logical switch port DHCP options, the kubevirt DHCP server is deactivated.

Reviewer: send DHCP replies advertising the allocated IP address to the guest VM (via OVN-Kubernetes DHCP options configured for the logical switch ports).
```suggestion
- Deliver IP address at VMs configuring logical switch port DHCP options, the kubevirt DHCP server is deactivated.
```

... or something like that.

The kubevirt DHCP part, while interesting, has nothing to do with OVN.Kubernetes, right ?

Author: > The kubevirt DHCP part, while interesting, has nothing to do with OVN.Kubernetes, right ?

Right, is just to understand that also KubeVirt is using DHCP too to offer the ip address, so reader see that this feature copy how is done at KubeVirt.

- A point to point routing is used so one node's subnet IP can be routed from different node

> - The VM's gateway IP and MAC are independent of the node they are running on using proxy arp

Reviewer: ```suggestion
- The VM's gateway IP and MAC are independent of the node they are running on using proxy ARP
```

**Point to point routing:**

> When the VM is running at a node that do not "own" it's IP address, like what happend

Reviewer: ```suggestion
When the VM is running at a node that does not "own" it's IP address, like what happend
```

after a live migration, do point to point routing with a policy for outbound traffic and a static route for
inbound. By doing this, the VM can live migrate to different node and keep previous 
addresses (IP / MAC), thus preserving n/s and e/w communication, and ensuring traffic
goes over the node where the VM is running. The latter reduces inter node communication.

If the VM is going back to the node that "owns" the ip, those static routes and 
policies should be reconciled (deleted).

```text
       ┌───────────────────────────┐
       │     ovn_cluster_router    │
┌───────────────────────────┐   ┌────────────────────────────┐
│ static route (ingress)    │───│ policy  (egress)           │
│ prefix: 10.244.0.8        │   │ match: ip4.src==10.244.0.8 │
│ nexthop: 10.244.0.8       │   │ action: reroute            │
│ output-port: rtos-node1   │   │ nexthop: 10.64.0.2         │
│ policy: dst-ip            │   │                            │  
└───────────────────────────┘   └────────────────────────────┘ 

> ```

Reviewer: while at it, might as well differentiate these: what's each entity purpose.

to route vm ingress traffic over the transit switch port address

If the transit switch port address is 10.64.0.3, those below are the static
route added to the topology when running with multiple zones:

Local zone:
       ┌───────────────────────────┐
       │     ovn_cluster_router    │
┌───────────────────────────┐      │
│ static route (dummy)      │──────┘
│ prefix: 10.244.0.0/24     │
│ nexthop: 10.64.0.3        │
│ policy: src-ip            │
└───────────────────────────┘

Remote zone:
       ┌───────────────────────────┐
       │     ovn_cluster_router    │
┌───────────────────────────┐      │
│ static route (ingress)    │──────┘
│ prefix: 10.244.0.8        │
│ nexthop: 10.64.0.3        │
│ policy: dst-ip            │
└───────────────────────────┘

**Nodes logical switch ports:**

The "router" logical switch port per node need to activate "arp_proxy" option
with a fixed MAC and VMs gateway so VM neighbor cache is consistent 
after live migration also cluster wide pod subnet need to be added so VM can 
ping IP's from same subnet after live migration, since arp proxy has less 
priority than than logical switch ARP flows there is no problem.

> 

Reviewer: please rephrase this one.

from, this means that the IPAM node subnet should go back to the pool but since
the migrated VMs contains IPs from it those IPs should reserved in case the 
subnet is assigned to a new node.

On that case before assigning the subnet to the node the VMs ips need to be 
reserved so they don't get assigned to new pods

Another scenario is ovn-kubernetes pods restarting after live migration, on 
that case ovnkube-master should discover to what IP pool the VM belongs.

To cover all this scenarios the VM's pods are labeled with `k8s.ovn.org/original-switch-name`
with the switch name where the VM was created since is the key for the IP pool to allocate IP, 
this label is copied during live migration to the new virt-launcher pod runnig
at the migration target node

#### Detecting migratable VMs and changing point to point routes

To detect that a pod is migratable KubevirtVm the annotation `kubevirt.io/allow-pod-bridge-network-live-migration`
has to be present and also the label `vm.kubevirt.io/name=vm1`, on that case
the point to point routes will be created after live migration and also a 
DHCPOptions will be configured to serve the ip configuration to the VM

> During live migration there is source and target virt-launcher pod's with different names, the ovn-k pod's controller use `CreationTimestamp` and `kubevirt.io/vm=vm1`
> to differentiate between them, then it watch the following annotation and label 
> at target pod `kubevirt.io/nodeName` and `kubevirt.io/migration-target-start-timestamp`:
> - The `kubevirt.io/nodeName` is set after the VM finishes live migrating or when it becomes ready.
> - The `kubevirt.io/migration-target-start-timestamp` is set when live migration has not finished but migration-target pod is ready to receive traffic (this happens at post-copy live migration, where migration is taking too long).

Reviewer: ```suggestion
During live migration there are two virt-launcher pods with different names for the same VM (source and target). The ovn-k pod controller uses `CreationTimestamp` and `kubevirt.io/vm=vm1`
to differentiate between them; it then watches the target pod `kubevirt.io/nodeName` and `kubevirt.io/migration-target-start-timestamp` annotation and label, with the following intent:
- The `kubevirt.io/nodeName` is set after the VM finishes live migrating or when it becomes ready.
- The `kubevirt.io/migration-target-start-timestamp` is set when live migration has not finished but migration-target pod is ready to receive traffic (this happens at post-copy live migration, where migration is taking too long).
```

... honestly I'm not that sure what I'm suggesting reads that better ...

Author: I think it does, I will take the changes.

The point to point routing cleanup (remove of static route and policy) will be done at two cases:
- VM is deleted.
- VM is live migrated back to the node that owns its IP.

> - ovn-kubernetes controllers are restarted (sync virtual machines)

Reviewer: I don't understand this item; it I restart the ovnkube master, will I remove the routes / policies for the affected nodes? The way it is written makes me believe that is what happens.

Author: It will remove "stale", I will add that to the last bullet.
